startup.py

```diff
@@ -23,83 +23,153 @@ class PhotoshopLauncher(SoftwareLauncher):
-                              like "2017", "6.3v7" or "1.2.3.52".
         :returns: List of :class:`SoftwareVersion` instances
         """
 
-        icon_path = os.path.join(self.disk_location, "resources", "ps_2017_icon_256.png")
+        software_versions = []
 
-        if sys.platform == "darwin":
-            # Default installs are located here:
-            # /Applications/Adobe Photoshop CC 2017/Adobe Photoshop CC 2017.app
-            glob_pattern = "/Applications/Adobe Photoshop CC */Adobe Photoshop CC *.app"
-            version_regex = re.compile(
-                "^/Applications/Adobe Photoshop CC (.+)/Adobe Photoshop CC (.+)\.app$",
-                re.IGNORECASE
-            )
+        for sw_version in self._find_software_versions():
+            if self.is_version_supported(sw_version):
+                self.logger.debug("Accepting %s", sw_version)
+                software_versions.append(sw_version)
+            else:
+                self.logger.debug("Rejecting %s", sw_version)
+                continue
 
-        elif sys.platform == "win32":
-            # Default installs are located here:
-            # C:\program files\Adobe\Adobe Photoshop CC 2017\Photoshop.exe
-            glob_pattern = r"C:\Program Files\Adobe\Adobe Photoshop CC *\Photoshop.exe"
-            version_regex = re.compile(
-                r"^C:\\Program Files\\Adobe\\Adobe Photoshop CC (.+)\\Photoshop.exe$",
-                re.IGNORECASE
-            )
+        return software_versions
+
+    def _find_software_versions(self):
+        """
+        Scan the filesystem for all photoshop executables.
+
+        :return: A list of :class:`SoftwareVersion` objects.
+        """
+
+        self.logger.debug("Scanning for photoshop executables...")
 
-        else:
+        # use the bundled icon
+        icon_path = os.path.join(
+            self.disk_location,
+            "resources",
+            "ps_2017_icon_256.png"
+        )
+        self.logger.debug("Using icon path: %s" % (icon_path,))
+
+        if sys.platform not in self.EXECUTABLE_MATCH_TEMPLATES:
             self.logger.debug("Photoshop not supported on this platform.")
             return []
 
-        self.logger.debug("Scanning for photoshop installations in '%s'..." % glob_pattern)
-        paths = glob.glob(glob_pattern)
+        # all the executable templates for the current OS
+        match_templates = self.EXECUTABLE_MATCH_TEMPLATES[sys.platform]
 
-        software_versions = []
-        for path in paths:
-            # extract version number
-            self.logger.debug("Found photoshop install in '%s'" % path)
-
-            match = version_regex.match(path)
-            if match:
-                # extract first group to get version string
-                dcc_version = match.group(1)
-                self.logger.debug("This is version '%s'" % dcc_version)
-
-                # see if we have a version filter
-                if versions and dcc_version not in versions:
```

Comment:
I really just wanted to see if this same logic that worked in nuke and houdini worked for PS. And it does. We don't have to keep this, but it does indicate that there are patterns and room for centralization. I personally think the abstraction of the sw paths is much preferable to hardcoding in the logic.